google-cloud-clients/google-cloud-contrib/google-cloud-spanner-jdbc/src/main/java/com/google/cloud/spanner/jdbc/SpannerPool.java

```diff
@@ -0,0 +1,407 @@
+    this.closeSpannerAfterMillisecondsUnused = closeSpannerAfterMillisecondsUnused;
+  }
+
+  /**
+   * Gets a Spanner object for a connection with the properties specified in the {@link
+   * ConnectionOptions} object. The {@link SpannerPool} will manage a pool of opened Spanner objects
+   * for the different connections, and reuse Spanner objects whenever possible. Spanner objects
+   * will also be closed down when the application is closing.
+   *
+   * @param options The specification of the Spanner database to connect to.
+   * @param connection The {@link ConnectionImpl} that will be created. This {@link ConnectionImpl}
+   *     will be tracked by the pool to know when a {@link Spanner} object can be closed.
+   * @return an opened {@link Spanner} object that can be used by a connection to communicate with
+   *     the Spanner database.
+   */
+  Spanner getSpanner(ConnectionOptions options, ConnectionImpl connection) {
+    Preconditions.checkNotNull(options);
+    Preconditions.checkNotNull(connection);
+    SpannerPoolKey key = SpannerPoolKey.of(options);
+    Spanner spanner;
+    synchronized (this) {
+      if (!initialized) {
+        initialize();
+      }
+      if (spanners.containsKey(key)) {
+        spanner = spanners.get(key);
+      } else {
+        spanner = createSpanner(key);
+        spanners.put(key, spanner);
+      }
+      List<ConnectionImpl> registeredConnectionsForSpanner = connections.get(key);
+      if (registeredConnectionsForSpanner == null) {
+        registeredConnectionsForSpanner = new ArrayList<>();
+        connections.put(key, registeredConnectionsForSpanner);
+      }
+      registeredConnectionsForSpanner.add(connection);
+      lastConnectionClosedAt.remove(key);
+      return spanner;
+    }
+  }
+
+  private void initialize() {
+    shutdownThread = new Thread(new CloseSpannerRunnable(), "SpannerPool shutdown hook");
+    Runtime.getRuntime().addShutdownHook(shutdownThread);
+    if (this.closeSpannerAfterMillisecondsUnused > 0) {
+      this.closerService =
+          Executors.newSingleThreadScheduledExecutor(
+              new ThreadFactory() {
+                @Override
+                public Thread newThread(Runnable r) {
+                  Thread thread = new Thread(r, "close-unused-spanners-worker");
+                  thread.setDaemon(true);
+                  return thread;
+                }
+              });
+      this.closerService.scheduleAtFixedRate(
+          new CloseUnusedSpannersRunnable(),
+          this.closeSpannerAfterMillisecondsUnused,
+          this.closeSpannerAfterMillisecondsUnused,
+          TimeUnit.MILLISECONDS);
+    }
+    initialized = true;
+  }
+
+  @SuppressWarnings("rawtypes")
+  @VisibleForTesting
+  Spanner createSpanner(SpannerPoolKey key) {
+    SpannerOptions.Builder builder = SpannerOptions.newBuilder();
+    builder
+        .setClientLibToken(MoreObjects.firstNonNull(key.userAgent, JdbcDriver.getClientLibToken()))
```

Comment:
Here is how spring-cloud-gcp does it: https://github.com/spring-cloud/spring-cloud-gcp/blob/master/spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/spanner/GcpSpannerAutoConfiguration.java#L126. I'd prefer this approach so we don't need to add the setter in the builder.

Reply:
That's something that we would need to discuss with the Spanner team in that case. The header used by spring-cloud-gcp sets a 'User-Agent' header, the client lib token of the JDBC driver sets the ApiClientHeader ('x-goog-api-client'). My initial solution was also what spring-cloud-gcp did, but it didn't show up in the monitoring that the Spanner team had. The ApiClientHeader did, which is why this solution was chosen, and the monitoring is now built around that.

That being said, I looked a little bit more into the spring-cloud-gcp solution and compared that with the original solution for the JDBC driver, and I noticed one important difference: spring-cloud-gcp sets a 'User-Agent' header, the JDBC driver used a 'user-agent' header, so that could be the reason it didn't show up.